lib/elsa/group/manager/worker_manager.ex

```diff
@@ -1,33 +1,61 @@
 defmodule Elsa.Group.Manager.WorkerManager do
+  @moduledoc """
+  Provides functions to encapsulate the management of worker
+  processes by the consumer group manager.
+  """
   import Record, only: [defrecord: 2, extract: 2]
   import Elsa.Group.Supervisor, only: [registry: 1]
 
   defrecord :brod_received_assignment, extract(:brod_received_assignment, from_lib: "brod/include/brod.hrl")
 
   defmodule WorkerState do
+    @moduledoc """
+    Tracks the running state of the worker process from the perspective of the group manager.
+    """
     defstruct [:pid, :ref, :generation_id, :topic, :partition, :latest_offset]
   end
 
+  @doc """
+  Retrieve the generation id, used in tracking assignments of workers to topic/partition,
+  from the worker state map.
+  """
+  @spec get_generation_id(map(), String.t(), integer()) :: integer()
   def get_generation_id(workers, topic, partition) do
     Map.get(workers, {topic, partition})
     |> Map.get(:generation_id)
   end
 
+  @doc """
+  Update the current offset for a given worker with respect to messages consumed
+  from its topic/partition.
+  """
+  @spec update_offset(map(), String.t(), integer(), integer()) :: map() | no_return()
   def update_offset(workers, topic, partition, offset) do
     Map.update!(workers, {topic, partition}, fn worker -> %{worker | latest_offset: offset + 1} end)
   end
 
+  @doc """
+  Iterate over all workers managed by the group manager and issue the unsubscribe call
+  to disengage from the topic/partition and shut down gracefully.
+  """
+  @spec stop_all_workers(map()) :: map()
   def stop_all_workers(workers) do
     workers
     |> Map.values()
     |> Enum.each(fn worker ->
       Process.demonitor(worker.ref)
-      GenServer.call(worker.pid, :unsubscribe)
+      Elsa.Group.Worker.unsubscribe(worker.pid)
     end)
 
     %{}
   end
 
+  @doc """
+  Restart the specified worker from the manager state. Retrieve the latest recorded
+  offset and pass it to the new worker to pick up where the previous left off if it
+  has been recorded.
+  """
+  @spec restart_worker(map(), reference(), struct()) :: map()
   def restart_worker(workers, ref, %Elsa.Group.Manager.State{} = state) do
     worker = get_by_ref(workers, ref)
     new_offset = if worker.latest_offset == :undefined, do: :undefined, else: worker.latest_offset
```

Comment:
Why is this if necessary? Aren't we assigning `new_offset` to `worker.latest_offset` either way? At one point were we incrementing the offset if it wasn't undefined?

Reply:
good catch, I am going to merge this PR and fix in another one.